File: leetcode-js/323-number-of-connected-components-in-an-undirected-graph.js
/**
 * @param {number} n
 * @param {number[][]} edges
 * @return {number}
 */
var countComponents = function(n, edges) {
    const roots = [];

    for (let i = 0; i < n; i += 1) {
        roots[i] = i;
    }

    for (const [node1, node2] of edges) {
        const root1 = find(roots, node1);
        const root2 = find(roots, node2);
        if (root1 !== root2) {
            roots[root1] = root2;
            n -= 1;
        }
    }

    return n;
};

function find(roots, node) {
    while (roots[node] !== node) {
        roots[node] = roots[roots[node]];
        node = roots[node];
    }
    return node;
}

File: leetcode-java/323-number-of-connected-components-in-an-undirected-graph.java
public class Solution {
    public int countComponents(int n, int[][] edges) {
        int[] roots = new int[n];
        for (int i = 0; i < n; i++) {
            roots[i] = i;
        }
        for (int[] edge : edges) {
            int a = find(roots, edge[0]);
            int b = find(roots, edge[1]);
            if (a != b) {
                n--;
                roots[a] = b;
            }
        }
        return n;
    }

    public int find(int[] roots, int a) {
        while (roots[a] != a) {
            roots[a] = roots[roots[a]];
            a = roots[a];
        }
        return a;
    }
}
